### ISSUE TITLE
libshiboken missing when pyside2 is in zip_include_packages

### ISSUE THREAD

**Issue by konserw** [marcelotduarte/cx_Freeze#665]
Hi!
I'm re-reporting my issue with `libshiboken dll missing` as it still is reproducible for me even with CX_freeze with #640 merged, given pyside2 and shiboken are in `zip_include_packages`.
Rationale for adding  pyside2 and shiboken to  `zip_include_packages` is reducing size of build from ~330MB to ~20MB.
I've created MR #664 with pyside sample that is reproducing my problem with PySide2 5.14.2

My original post from #638 for reference:

> 
> It's probably this: https://bugreports.qt.io/browse/PYSIDE-1257
> PySide2 5.14.2 introduced this new method:
> 
> def _additional_dll_directories(package_dir):
>     # Find shiboken2 relative to the package directory.
>     root = os.path.dirname(package_dir)
>     shiboken2 = os.path.join(root, 'shiboken2')
>     if os.path.isdir(shiboken2): # Standard case, only shiboken2 is needed
>         return [shiboken2]
>     # The below code is for the build process when generate_pyi.py
>     # is executed in the build directory. We need libpyside and Qt in addition.
>     shiboken2 = os.path.join(os.path.dirname(root), 'shiboken2', 'libshiboken')
>     if not os.path.isdir(shiboken2):
>         raise ImportError(shiboken2 + ' does not exist')
>     result = [shiboken2, os.path.join(root, 'libpyside')]
>     for path in os.environ.get('PATH').split(';'):
>         if path:
>              if os.path.exists(os.path.join(path, 'qmake.exe')):
>                  result.append(path)
>                  break
>     return result
> 
> Which breaks cx_forzen apps (at least on windows). Basically in normal flow (not frozen) if os.path.isdir(shiboken2): evaluates to True, so method returns in next line and all is good; but if in frozen case it evaluates to False (as shiboken dir is not where it looks for it), so it looks for linux lib or something, so it rises exception as tha can't be found either.
> 
> It's regression in PySide code, but I'm not sure if it have to be fixed there, or can be worked around here. Please cooperate with them on this :)
> 
> I hope this is helpful.

In mentioned qt bug there is comment that:

> This is definitely an issue with Python 3.8 and the additional_dll_directories method.  In a frozen application the root variable points to library.zip.  This method assumes that root points to a normal directory.

I'm not sure if this is something that (or should) can be fixed by CX_Freeze. But maybe you could advice how to proceed. I.e. frozen application works OK when pyside2 is included "normally" not in zip packages, but it leads to over 300MB of bloat in package

**Comment by piti-diablotin** [comment]
Hi,

I am having a similar issue I guess.
First, here is my list of installed packages with python 3.8.3 for windows

      PySide2 5.15.0
      shiboken2 5.15.0
      cx-freeze 6.1

I can run a simple PySide2 code (like in #664) but I cannot freeze the app.
I tried with and without `zip_include_packages` but nothing work, even if the error seems different. It is always a `shiboken2` error.

Any idea to resolve the situation ?

Thanks


![2020-06-14_17h35_41](https://user-images.githubusercontent.com/27724801/84597732-4245f680-ae66-11ea-8521-5f66ab53c2f4.png)
![2020-06-14_17h36_31](https://user-images.githubusercontent.com/27724801/84597750-5853b700-ae66-11ea-80f2-0346ec4b81da.png)

**Comment by konserw** [comment]
@piti-diablotin simplest workaround is installing pyside2 5.14.1 and that's what I'm using for now. https://bugreports.qt.io/browse/PYSIDE-1257 is already fixed, but we need to wait for fix to be published as part of some pyside2 release

**Comment by piti-diablotin** [comment]
Thanks it works now !
However, not related to this issue particularly, how do you select what you want to include of PySide2 and not the full package ? I tried `excludes` but does not work. 
PySide2 is quite heavy for just a basic QWidget app...
Thanks

**Comment by konserw** [comment]
check sample from my PR: https://github.com/anthony-tuininga/cx_Freeze/pull/664 there is set of option I use for my app:

```
options = {
    'build_exe': {
        "include_files": [os.path.join(plugins_path, "platforms")],  # additional plugins needed by qt at runtime
        "zip_include_packages": ["PySide2", "shiboken2", "encodings"],  # reduce size of packages that are used
        "excludes": ["tkinter", "unittest", "email", "http", "xml", "pydoc", "pdb"],  # exclude packages that are not really needed
    }
}
````

**Comment by piti-diablotin** [comment]
Thanks. I try to work with this example but when I deploy the package I then get
```
This application failed to start because no Qt platform plugin could be initialized. \
Reinstalling the application may fix this problem.
```
Did I miss something ?
Without the `zip_include_packages` everything is fine ...

**Comment by konserw** [comment]
Yep - the line with platform dir include

pon., 22 cze 2020, 21:45 użytkownik piti-diablotin <notifications@github.com>
napisał:

**Comment by piti-diablotin** [comment]
Oh yes sorry, now it works, but the theme is not the default windows' one but the old fashion win98 style ... Any reason ? 
Thanks for you very appreciated help !

EDIT:I build also the package on linux, I am not sure to know how to make the "dist" file if there is one to do ?
Just copy/pasting the build directory to a linux system I got 
```
ImportError: libpyside2.cpython-36m-x86_64-linux-gnu.so.5.11: cannot open shared object file: No such file or directory
```
which I included then by hand but still does not work.

**Comment by piti-diablotin** [comment]
I figured out the issue I had on windows. I missed the styles and imageformats in the `include_files` in addition to platforms!
Issue on linux is remaining though :smile:

**Comment by rtrobin** [comment]
Hi guys. @konserw @piti-diablotin. I met same issue with PySide2==5.15.0. I now downgraded it to 5.14.1.

I'm writing an app on Win10, using python==3.8.5 which ms store offers. My app uses `PySide2.QtUiTools.QUiLoader` to load an `.ui` file. By using the setup.py script @konserw provides, it freezes the script without error. But the app cannot run due to the loss of `PySide2.QtXml`.

![image](https://user-images.githubusercontent.com/8224989/90978982-438d5100-e584-11ea-9ec2-b6f7dba580d3.png)

In my code, I add `import PySide2.QtXml` to import it explicitly. Though it could build and run without error now, it shows like old win98 style. Do you guys know how to fix it? Thanks in advance.

By the way, the output directory is shown as below. Just to make sure I did things right. 

![image](https://user-images.githubusercontent.com/8224989/90979092-30c74c00-e585-11ea-877e-22cf3efe0faa.png)

**Comment by piti-diablotin** [comment]
Yes I figured out the solution to the win98 old style.You miss the `styles` and `imageformats` in the `include_files`
My `setup.py` contains
``` python
 if sys.platform == 'win32':
     base = 'Win32GUI'
     include_files += [
         os.path.join(plugins_path, "platforms"),
         os.path.join(plugins_path, "styles"),
         os.path.join(plugins_path, "imageformats"),
         "win","C:/Program Files/Python38/vcruntime140.dll"]
```
This is what I had to do to make it work on windows
I hope this helps

**Comment by rtrobin** [comment]
@piti-diablotin Thank you! Now I realize I didn't understand your previous comment. 

> I figured out the issue I had on windows. I missed the styles and imageformats in the `include_files` in addition to platforms!
> Issue on linux is remaining though 😄

**Comment by maicol07** [comment]
It seems that 5.14.2.3 fixes the issue: pyside/pyside2-setup@88af089, although a release for 5.15.x is missing...

**Comment by kopaw** [comment]
> 
> 
> It seems that 5.14.2.3 fixes the issue: [pyside/pyside2-setup@88af089](https://github.com/pyside/pyside2-setup/commit/88af089), although a release for 5.15.x is missing...

I've just tested Pyside2 5.15.2 x64, where this fix is included with Python 3.7.9, 3.8.6, and 3.9.0 (all x64) on Windows 10 20H2 (19042.630) and cx_Freeze 6.3, but the issue still persists.

**Comment by marcelotduarte** [comment]
[cx_Freeze 6.10](https://github.com/marcelotduarte/cx_Freeze/releases/tag/6.10) has just been released.
